index/scorch/optimize_knn.go

```diff
@@ -57,22 +69,26 @@ func (o *OptimizeVR) Finish() error {
 					wg.Done()
 				}()
 				for field, vrs := range o.vrs {
-					searchVectorIndex, closeVectorIndex, err := segment.InterpretVectorIndex(field)
+					vecIndex, err := segment.InterpretVectorIndex(field)
 					if err != nil {
 						errorsM.Lock()
 						errors = append(errors, err)
 						errorsM.Unlock()
 						return
 					}
+
+					// update the vector index size as a meta value in the segment snapshot
+					vectorIndexSize := vecIndex.Size()
+					seg.cachedMeta.updateMeta(field, vectorIndexSize)
```

Comment:
I'm not sure I follow this. Isn't it here we want to prevent a Search() if the app_herder deems it so. Or if i am interpreting this right - you want subsequent access to these segments to be stopped on this account?

Reply:
Basically the thing is that if I put the check at this point, then I _must_ read the index (which is expensive) get the vector index and its size and then subject it to the app_herder check. It also comes with the drawback that the locks in the app_herder code would serialize the deployed goroutines over here which parallelizes the things. Another drawback is that now the app_herder check would be deployed for each segment.

So, in order to avoid too many redundant app_herder calls (its sufficient to invoke it once for every searcher creation since that's what we are trying to limit) and also to prevent the reading of the index from the buffer altogether in case of app_herder rejection, I pushed the check to one level above (i.e. at the VectorOptimize() function which is invoked once for every searcher).

In order to achieve this - the approach I have taken is that the very first thread which hits those segments and loads index now just records the size of the index as a meta data in the segment snapshot - and the reason why I used this struct is because one the context is established that we are storing something specific to a segment and the other reason is I followed the technique of how cachedDocs is populated and used for doc value retrieval (and i think that codepath is hit when the docvalue option is not enabled in the type mappings). The subsequent threads 
which corresponding to different queries which create a different OptimizeVR object altogether now consult the underlying segment snapshot which is now populated with the vector index size and then perform an app_herder check hence making sure that we are invoking this check only once for every searcher and not len(segments) times as earlier.

In the Finish() function we basically invoke the end callback as a defer logic.

Reply:
Ok, looks like it was what I thought. Let me do one more review in some time.